docs/source/design/dataset.md

```diff
@@ -0,0 +1,331 @@
+
+### Design
+
+#### General workflow for dataset load and processing
+Before introducing the Olive `DatasetContainer` interface, let's first outline the general workflow for loading and processing datasets in machine learning.
+![figure](../images/dataset-flow.png)
+
+Based on the above workflow, we can see that dataset loading and processing can be divided into three steps:
+
+1. User to prepare the dataset with given format before call Olive: `[data, annotation(optional), meta_data(optional)]`.
+    - `data`: The input data for the model to process. This can be a single tensor, text, file, or a list of tensors, texts, or files.
+    - `annotation`: The label for the data. This can be a single tensor, text, file, or a list of tensors, texts, or files.
+    - `meta_data`: The metadata for the data. This includes information that can help with understanding the structure or parsing of the current data.
+2. After preparing the data, users can implement their own methods to load the dataset and obtain the dataloader.
+The steps would be like:
+    - `dataset` to construct the dataset.
+    - `pre_process` to format the dataset and make it ready for model to consume. Could be optional if the dataset is ready for model inference. Noted that the data after `dataset` and `pre_process` should be return `(data, label)` with `__getitem__` methods.
+    - `post_precess` to format the model output to the format which can be used for evaluation. Could be optional.
+    - `data_loader` to generate the dataloader with given batch_size/sample_ratio.
+After these, the DataContainer will generate the dataloader with given batch_size/sample_ratio which can be used for model inference. User can also overwrite it with their own version.
+
+3. After the dataloader is ready, user can call Olive to run the pass or evaluation.
+
+
+#### Olive DataContainer interface
+Based on above workflow, we introduce the Olive `DataContainer` interface which
+1. unifies the dataset interface for different dataset, models and tasks.
+2. gives the chance for Olive to implement the build-in `dataset`, `pre_process` and `post_process` and `dataloader`.
+3. simplifies the user experience to implement their own `dataset`, `pre_process` and `post_process` and `dataloader`.
+4. simplifies the user experience for some popular dataset models and tasks.
+
+Then the design for `DataContainer` interface will be like:
+![figure](../images/datacontainer_example.png)
+
+1. There will be build-in DataContainers for different datasets, models and tasks, which implement the Olive `DataContainer` interface, like `HuggingfaceDataContainer`, `VideoDataContainer`, `VoiceDataContainer` and more.
+Each build-in DataContainer uniquely corresponds to a template of config like:
+    ```json
+    {
+        "data_container": {
+            "type": "BaseContainer", // Olive BaseContainer interface, can be other available DataContainers like HuggingfaceDataContainer, VideoDataContainer...
```

Comment:
We should not expose data_container to users for standard HF use cases.

Reply:
Updated. Will update whisper case after this PR is merged.